src/server/frontend_wayland/wl_surface.h

```diff
@@ -198,6 +201,9 @@ class WlSurface : public wayland::Surface
     NullWlSurfaceRole null_role;
     WlSurfaceRole* role;
     std::vector<WlSubsurface*> children; // ordering is from bottom to top
+    ptrdiff_t parent_z_index{0}; // index in children where parent surface renders (subsurfaces before this are below parent)
+    std::optional<std::list<WlSubsurface*>> pending_children_order; // pending z-order changes (using list for simpler reordering)
+    std::list<WlSubsurface*>::iterator pending_parent_position; // iterator to position where parent renders in pending_children_order
```

Comment:
This feels like it should be in `WlSurfaceState` (specifically, the `pending` member)?